tools/python/stylesheet/color_matrix_builder.py

```diff
@@ -0,0 +1,244 @@
+            drules.ParseFromString(infile.read())
+        return drules
+
+
+    # READING:
+    def reader_factory(self, index):
+        def reader(sub_drules, index_list):
+            for key, _, color in self.color_key_generator(sub_drules, index_list):
+                self.all_elements[index][key] = color
+        return reader
+
+
+    def parse_all_bins(self):
+        print("Parsing the bin files")
+        self.foreach_path(lambda index: self.parse_bin(index))
+
+
+    def parse_bin(self, index):
+        self.recursive_iterator(self.drules[index], "", index, self.reader_factory(index))
+
+
+    # CALCULATE COLOR MATRIX
+    def calculate_color_matrix(self):
+        self.normalize_all_keys()
+        all_keys = set.union(*map(lambda s: set(s[1]), self.all_elements.iteritems()))
+        colors_table = defaultdict(list)  # {key: [color_0, color_1, -1 if element absent in style 2, ...]}
+
+        for key in all_keys:
+            for path_index in xrange(len(self.paths)):
+                colors_table[key].append(self.all_elements[path_index][key])
+
+        for key, colors in colors_table.iteritems():
+            self.transformation_table[tuple(colors)].append(key)
+
+
+    # WRITE COLOR MATRICES
+    def write_all_matrices(self):
+        print("Writing matrices")
+        self.foreach_path(lambda index: self.write_matrix(index))
+
+
+    def write_matrix(self, index):
+        colors = map(lambda x: x[index], self.transformation_table)
+
+        imgmk = ImageMaker(colors)
+        imgmk.make()
+        imgmk.save(self.new_filename(self.paths[index], "png"))
+        self.imgs.append(imgmk)
+
+
+    def new_filename(self, bin_path, postfix):
+        folder = path.dirname(bin_path)
+        style_name = self.calculate_style_type_postfix(bin_path)
+        return path.join(folder, "static_colors{}.{}".format(style_name, postfix))
+
+
+    def calculate_style_type_postfix(self, bin_path):
+        return path.basename(bin_path)[len("drules_proto"):-4]
+
+
+    # UPDATE BINS
+    def invert_transformation_table(self):
+        for element, (i, _) in zip(self.transformation_table.iteritems(), enumerate(self.transformation_table)):
+            for key in element[1]:
+                self.inverted_transformation_table[key] = i
+
+
+    def update_all_bins(self):
+        print("Updating bin files")
+        self.invert_transformation_table()
+        self.foreach_path(self.update_one_bin)
+
+
+    def update_one_bin(self, index):
+        self.recursive_iterator(self.drules[index], "", index, self.updater_factory(index))
+
+
+    def write_all_bins(self):
+        print("Writing bin files")
+        self.foreach_path(self.write_one_bin)
+
+
+    def write_one_bin(self, path_index):
+        new_bin_name = self.paths[path_index]
+        with open(new_bin_name, "wb") as outfile:
+            outfile.write(self.drules[path_index].SerializePartialToString())
+
+        with open(new_bin_name[:-4] + ".txt", "wb") as outfile:
```

Comment:
Part only after last dot will be replaced. You are right, it may be better to use https://docs.python.org/2/library/os.path.html#os.path.splitext